app/models/document.rb

```diff
@@ -104,12 +104,26 @@ def humanized_attributes
     end
   end
 
+  def self.extract_body_from_payload(payload)
```

Comment:
This method is quite brittle in that it will raise an error at runtime if the payload doesn't exactly match the structure it's expecting. Although this is fine, it might be worth failing fast (via `#fetch`) or raising a more helpful error if the structure is incorrect.